File: src/main/java/player/ki/DynamicEvaluation.java
package player.ki;

import board.ChessGameInterface;
import helper.ChessGameSupervisor;
import helper.Move;

import java.util.ArrayList;
import java.util.List;

/**
 * @author stephan
 */
public class DynamicEvaluation {
    final private int possibleMovesBufferSize = 40;

    private StaticEvaluationInterface strategy;
    private SearchTreePruner pruner;

    DynamicEvaluation(SearchTreePruner pruner, StaticEvaluationInterface strategy) {
        setEvaluationStrategy(strategy);
        setSearchTreePruner(pruner);
    }

    DynamicEvaluation() {
        this(new SimplePruner(), new StaticEvaluation());
    }

    public Evaluaded evaluateMove(final ChessGameInterface game, final Move move) {
        int depth = 0;
        final boolean forWhite = game.isWhiteTurn();
        boolean thisMove_hasHitFigure = game.hasHitFigure();
        boolean thisMove_isChess = game.isCheck(!forWhite);

        final ChessGameSupervisor interactiveSupervisor = game.suspendInteractiveSupervisor();
        int endoption = game.move(move);

        Evaluaded result;
        if (endoption == ChessGameInterface.NO_END) {
            final List<Move> minPossibleMovesBuffer = new ArrayList<Move>(possibleMovesBufferSize);
            result = getMin(game, forWhite, depth, thisMove_isChess, thisMove_hasHitFigure, minPossibleMovesBuffer);
        } else if (endoption == ChessGameInterface.MATT) {
            result = Evaluaded.getThisComputerPlayerIsMatt(depth + 1);
        } else {
            result = Evaluaded.DRAW;
        }

        game.useSupervisor(interactiveSupervisor);
        game.undo();
        return result;
    }

    private Evaluaded getMin(final ChessGameInterface game,
                         final boolean forWhite,
                         final int depth,
                         final boolean lastMove_isChess,
                         final boolean lastMove_hasHitFigure,
                         final List<Move> minPossibleMovesBuffer) {
        boolean thisMove_hasHitFigure = game.hasHitFigure();
        boolean thisMove_isChess = game.isCheck(!forWhite);

        if (pruner.stopMinDynamicEvaluation(depth,
                thisMove_isChess,
                thisMove_hasHitFigure,
                lastMove_isChess,
                lastMove_hasHitFigure)) {
            return strategy.evaluate(game, forWhite);
        }

        Evaluaded minValue = Evaluaded.INITAL;

        minPossibleMovesBuffer.clear();
        game.getPossibleMoves(minPossibleMovesBuffer);
        final List<Move> maxPossibleMovesBuffer = new ArrayList<Move>(possibleMovesBufferSize);

        for (Move move : minPossibleMovesBuffer) {
            Evaluaded tempValue;

            assert (game.isFreeArea(move.to) || !game.getFigure(move.to).isKing())
                    : "getMin:" +
                    game.getFigure(move.from).getClass().getName() +
                    " hits King white Move " +
                    move.toString();

            int endoption = game.move(move);

            if (endoption == ChessGameInterface.NO_END) {
                tempValue = getMax(game,
                        forWhite,
                        depth,
                        thisMove_isChess,
                        thisMove_hasHitFigure,
                        maxPossibleMovesBuffer);
            } else if (endoption == ChessGameInterface.MATT) {
                game.undo();
                return Evaluaded.getOtherPlayerIsMatt(depth + 1);
            } else {
                tempValue = Evaluaded.DRAW;
            }

            game.undo();

            if (minValue == Evaluaded.INITAL || minValue.compareTo(tempValue)>0) {
                minValue = tempValue;
            }
        }

        assert minValue != Evaluaded.INITAL
                : "no minimum found";

        return minValue;
    }

    private Evaluaded getMax(final ChessGameInterface game,
                         final boolean forWhite,
                         int depth,
                         final boolean lastMove_isChess,
                         final boolean lastMove_hasHitFigure,
                         final List<Move> maxPossibleMovesBuffer) {
        depth++;

        boolean thisMove_hasHitFigure = game.hasHitFigure();
        boolean thisMove_isChess = game.isCheck(forWhite);
        if (pruner.stopMaxDynamicEvaluation(depth,
                thisMove_isChess,
                thisMove_hasHitFigure,
                lastMove_isChess,
                lastMove_hasHitFigure)) {
            return strategy.evaluate(game, forWhite);
        }

        Evaluaded maxValue = Evaluaded.INITAL;

        maxPossibleMovesBuffer.clear();
        game.getPossibleMoves(maxPossibleMovesBuffer);
        final List<Move> minPossibleMovesBuffer = new ArrayList<Move>(possibleMovesBufferSize);

        for (Move move : maxPossibleMovesBuffer) {
            Evaluaded tempValue;

            assert (game.isFreeArea(move.to) || !game.getFigure(move.to).isKing())
                    : "getMax:" +
                    game.getFigure(move.from).getClass().getName() +
                    " hits King white Move " +
                    move.toString();

            int endoption = game.move(move);

            if (endoption == ChessGameInterface.NO_END) {
                tempValue = getMin(game,
                        forWhite,
                        depth,
                        thisMove_isChess,
                        thisMove_hasHitFigure,
                        minPossibleMovesBuffer);
            } else if (endoption == ChessGameInterface.MATT) {
                game.undo();
                return Evaluaded.getThisComputerPlayerIsMatt(depth + 1);
            } else {
                tempValue = Evaluaded.DRAW;
            }

            game.undo();

            if (maxValue == Evaluaded.INITAL || tempValue.compareTo(maxValue)>0) {
                maxValue = tempValue;
            }
        }

        assert maxValue != Evaluaded.INITAL
                : "no maximum found";

        return maxValue;
    }

    public void setEvaluationStrategy(StaticEvaluationInterface strategy) {
        this.strategy = strategy;
    }

    public void setSearchTreePruner(SearchTreePruner pruner) {
        this.pruner = pruner;
    }
}

File: src/main/kotlin/voidchess/player/ki/evaluation/DynamicEvaluation.kt
package voidchess.player.ki.evaluation

import voidchess.board.ChessGameInterface
import voidchess.board.getFigure
import voidchess.board.move.Move
import voidchess.board.move.MoveResult
import java.util.*
import java.util.Collections.emptyList

class DynamicEvaluation(var pruner: SearchTreePruner, var strategy: StaticEvaluationInterface) {

    constructor() : this(PrunerWithIrreversibleMoves(), StaticEvaluation)

    fun evaluateMove(game: ChessGameInterface, move: Move): Evaluated {
        val depth = 0
        val forWhite = game.isWhiteTurn
        val thisMoveHasHitFigure = game.hasHitFigure()
        val thisMoveIsChess = game.isCheck(!forWhite)

        val interactiveSupervisor = game.suspendInteractiveSupervisor()
        val endOption = game.move(move)

        val result = when {
            endOption === MoveResult.NO_END -> {
                getMin(game, forWhite, depth, thisMoveIsChess, thisMoveHasHitFigure)
            }
            endOption === MoveResult.CHECKMATE -> CheckmateOther(depth + 1)
            else -> Draw
        }

        game.useSupervisor(interactiveSupervisor)
        game.undo()
        return result
    }

    private fun getMin(game: ChessGameInterface,
                       forWhite: Boolean,
                       depth: Int,
                       lastMove_isChess: Boolean,
                       lastMove_hasHitFigure: Boolean
    ): Evaluated {
        val thisMoveHasHitFigure = game.hasHitFigure()
        val thisMoveIsChess = game.isCheck(!forWhite)

        val minPossibleMovesBuffer =
                when (pruner.continueMinDynamicEvaluationBy(depth,
                        thisMoveIsChess,
                        thisMoveHasHitFigure,
                        lastMove_isChess,
                        lastMove_hasHitFigure
                )) {
                    ContinueEvalBy.StaticEval -> emptyMoveList
                    ContinueEvalBy.AllMoves -> game.getAllMoves()
                    ContinueEvalBy.IrreversibleMoves -> game.getCriticalMoves()
                    ContinueEvalBy.TakingMoves -> game.getTakingMoves()
                }

        if(minPossibleMovesBuffer.isEmpty()) {
            return strategy.getPrimaryEvaluation(game, forWhite)
        }

        val primaryEvaluations = TreeSet<EvaluatedMove>()

        for (move in minPossibleMovesBuffer) {

            assert(game.isFreeArea(move.to) || !game.getFigure(move.to).isKing()) {
                "getMin: ${game.getFigureOrNull(move.from)} hits King white Move $move"
            }

            val endOption = game.move(move)

            val primaryEval = when {
                endOption === MoveResult.NO_END -> getMax(
                        game,
                        forWhite,
                        depth,
                        thisMoveIsChess,
                        thisMoveHasHitFigure
                )
                endOption === MoveResult.CHECKMATE -> {
                    game.undo()
                    return CheckmateSelf(depth + 1)
                }
                else -> Draw
            }

            primaryEvaluations.add(EvaluatedMove(move, primaryEval))

            game.undo()
        }

        //now byCode a secondaryEvaluation for the Moves with the smallest primary Evaluation
        val combinedEvaluations = TreeSet<Evaluated>()
        val minPrimary = primaryEvaluations.first().value

        for (evaluatedMove in primaryEvaluations) {
            val primaryEval = evaluatedMove.value
            if (primaryEval.isCloseToByPrimary(minPrimary)) {
                game.move(evaluatedMove.move)
                //add secondary Evaluation
                strategy.addSecondaryEvaluation(game, forWhite, primaryEval)
                combinedEvaluations.add(primaryEval)
                game.undo()
            } else {
                //the set is sorted so once an element is to big, all the rest are
                break
            }
        }

        //minValue
        return combinedEvaluations.first()
    }

    private fun getMax(game: ChessGameInterface,
                       forWhite: Boolean,
                       oldDepth: Int,
                       lastMove_isChess: Boolean,
                       lastMove_hasHitFigure: Boolean
    ): Evaluated {
        val depth = oldDepth + 1

        val thisMoveHasHitFigure = game.hasHitFigure()
        val thisMoveIsChess = game.isCheck(forWhite)

        val maxPossibleMovesBuffer =
                when (pruner.continueMaxDynamicEvaluationBy(depth,
                        thisMoveIsChess,
                        thisMoveHasHitFigure,
                        lastMove_isChess,
                        lastMove_hasHitFigure
                )) {
                    ContinueEvalBy.StaticEval -> emptyMoveList
                    ContinueEvalBy.AllMoves -> game.getAllMoves()
                    ContinueEvalBy.IrreversibleMoves -> game.getCriticalMoves()
                    ContinueEvalBy.TakingMoves -> game.getTakingMoves()
                }

        if(maxPossibleMovesBuffer.isEmpty()) {
            return strategy.getPrimaryEvaluation(game, forWhite)
        }

        val primaryEvaluations = TreeSet<EvaluatedMove>()

        for (move in maxPossibleMovesBuffer) {

            assert(game.isFreeArea(move.to) || !game.getFigure(move.to).isKing()) {
                "getMax: ${game.getFigureOrNull(move.from)} hits King white Move $move"
            }

            val endOption = game.move(move)

            val primaryEval = when {
                endOption === MoveResult.NO_END -> getMin(
                        game,
                        forWhite,
                        depth,
                        thisMoveIsChess,
                        thisMoveHasHitFigure
                )
                endOption === MoveResult.CHECKMATE -> {
                    game.undo()
                    return CheckmateOther(depth + 1)
                }
                else -> Draw
            }

            primaryEvaluations.add(EvaluatedMove(move, primaryEval))

            game.undo()
        }

        //now byCode a secondaryEvaluation for the Moves with the smallest primary Evaluation
        val combinedEvaluations = TreeSet<Evaluated>()
        val maxPrimary = primaryEvaluations.last().value

        for (evaluatedMove in primaryEvaluations.descendingSet()) {
            val primaryEval = evaluatedMove.value
            if (primaryEval.isCloseToByPrimary(maxPrimary)) {
                game.move(evaluatedMove.move)
                //add secondary Evaluation
                strategy.addSecondaryEvaluation(game, forWhite, primaryEval)
                combinedEvaluations.add(primaryEval)
                game.undo()
            } else {
                //the set is sorted in reverse so once an element is to small, all the rest are
                break
            }
        }

        //maxValue
        return combinedEvaluations.last()
    }

    companion object {
        private val emptyMoveList: List<Move> = emptyList<Move>()
    }
}